### PR TITLE
[JUJU-1635, JUJU-1782] Rewrite api/client/firewallrules unit tests to use gomock

### PR DESCRIPTION
**Pull request by cderici**
The unit tests for api/client/firewallrules now use go mock and not juju/testing. All unit tests should pass.

### CODE REVIEW

File: api/client/firewallrules/client_test.go [hunk 1]
```diff
@@ -4,140 +4,114 @@
...
-	client := firewallrules.NewClient(apiCaller)
+	args := params.FirewallRuleArgs{
+		Args: []params.FirewallRule{{
+			KnownService: "ssh",
+		}},
+	}
+	res := new(params.ErrorResults)
+	results := params.ErrorResults{
+		Results: []params.ErrorResult{{
+			Error: apiservererrors.ServerError(errors.New("fail"))}},
```
Comment (hmlanigan) [inline api/client/firewallrules/client_test.go:60]: this part will never be evaluated by the client. best to remove and not confusing the test intention.

File: api/client/firewallrules/client_test.go [hunk 2]
```diff
@@ -4,140 +4,114 @@
...
-
-	client := firewallrules.NewClient(apiCaller)
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	mockFacadeCaller := basemocks.NewMockFacadeCaller(ctrl)
+	client := firewallrules.NewClientFromCaller(mockFacadeCaller)
+
 	err := client.SetFirewallRule("foo", []string{"192.168.1.0/32"})
 	c.Assert(err, gc.ErrorMatches, `known service "foo" not valid`)
```
Comment (hmlanigan) [inline api/client/firewallrules/client_test.go:79]: suggested matching on error type of NotValid instead, then if the string changes, there is less to update.

Comment (hmlanigan) [inline api/client/firewallrules/client_test.go:79]: side note: this error is awkward for the user, how is a "known" service not valid. as a drive-by you could update the text of the error message in params.

File: api/client/firewallrules/client_test.go [hunk 3]
```diff
@@ -4,140 +4,114 @@
...
+	client := firewallrules.NewClientFromCaller(mockFacadeCaller)
+
+	ress, err := client.ListFirewallRules()
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(called, jc.IsTrue)
-	c.Assert(results, jc.DeepEquals, []params.FirewallRule{{
+	c.Assert(ress, jc.DeepEquals, []params.FirewallRule{{
 		KnownService:   params.SSHRule,
 		WhitelistCIDRS: []string{"192.168.1.0/32"},
 	}})
```
Comment (hmlanigan) [inline api/client/firewallrules/client_test.go:103]: suggest: c.Assert(ress, jc.DeepEquals, result.Rule[0])

**Review by hmlanigan — COMMENTED** [review]
Now that you have the hang of converting these tests to go mock. Suggest you look at potential improvements as well. Some of these are very old tests.

File: api/client/firewallrules/client_test.go [hunk 4]
```diff
@@ -4,140 +4,107 @@
...
-	client := firewallrules.NewClient(apiCaller)
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	mockFacadeCaller := basemocks.NewMockFacadeCaller(ctrl)
+	client := firewallrules.NewClientFromCaller(mockFacadeCaller)
+
 	err := client.SetFirewallRule("foo", []string{"192.168.1.0/32"})
-	c.Assert(err, gc.ErrorMatches, `known service "foo" not valid`)
+	errors.Is(err, errors.NotValid)
```
Comment (hmlanigan) [inline api/client/firewallrules/client_test.go:75]: The result needs to be evaluated.

An alternative method is c.Assert(err, jc.Satisfies, errors.IsNotValid)

**Review by hmlanigan — APPROVED** [review]
A tiny tweak is needed, then LGTM

File: api/client/firewallrules/client_test.go (hunk 4 above)
Comment (cderici) [reply api/client/firewallrules/client_test.go:75]: @hmlanigan By result, do you mean the error message? (cause there's no other result from `SetFirewallRules` it's running for a side effect.)

Also, the `errors.IsNotValid` says that it's being deprecated.

**Comment by cderici** [comment]
That failing test `secrets_test.go:627: SecretsSuite.TestListSecretRevisions` in [github-make-check-juju](https://jenkins.juju.canonical.com/job/github-make-check-juju/2460/) job is passing for me, might be a glitch. Gonna try to merge again, /merge
